### ISSUE TITLE
[YSQL] Inconsistent follower read results

### ISSUE THREAD

**Issue by jameshartig** [yugabyte/yugabyte-db#12855]
Jira Link: [DB-2617](https://yugabyte.atlassian.net/browse/DB-2617)
### Description

With the following Go code:

```go
package main

import (
	"context"
	"fmt"
	"time"

	"github.com/jackc/pgx/v4"
)

func main() {
	writeConn, err := pgx.Connect(context.Background(), "postgresql://yugabyte@127.0.0.1:5433/yugabyte")
	if err != nil {
		panic(err)
	}

	readConn, err := pgx.Connect(context.Background(), "postgresql://yugabyte@127.0.0.1:5433/yugabyte?yb_read_from_followers=true&yb_follower_read_staleness_ms=5000")
	if err != nil {
		panic(err)
	}

	_, err = writeConn.Exec(context.Background(), "CREATE TABLE IF NOT EXISTS t (t1 INT, PRIMARY KEY (t1))")
	if err != nil {
		panic(err)
	}
	_, err = writeConn.Exec(context.Background(), "TRUNCATE TABLE t")
	if err != nil {
		panic(err)
	}

	go func() {
		t := time.NewTicker(time.Second)
		i := 0
		for range t.C {
			i++
			_, err := writeConn.Exec(context.Background(), "INSERT INTO t (t1) VALUES ($1)", i)
			if err != nil {
				panic(err)
			}
			fmt.Println("set", i)
		}
	}()

	time.Sleep(10 * time.Second)
	t := time.NewTicker(time.Second)
	for range t.C {
		i := new(int)
		err := readConn.QueryRow(context.Background(), "/*+ Set(transaction_read_only true) */ SELECT MAX(t1) FROM t").Scan(&i)
		if err != nil {
			panic(err)
		}
		if i == nil {
			fmt.Println("get", "null")
		} else {
			fmt.Println("get", *i)
		}
	}
}
```

I get:

```
set 1
set 2
set 3
set 4
set 5
set 6
set 7
set 8
set 9
set 10
set 11
get 6
set 12
get 12
set 13
get 13
set 14
get 14
set 15
get 15
set 16
get 16
set 17
get 17
```

The logs for `get 6` were:
```
I0611 02:32:53.198423   159 tablet_service.cc:2067] Received Read RPC: tablet_id: "8e32c8a62c4b4167abcc9f862bed812f"
include_trace: false
consistency_level: CONSISTENT_PREFIX
propagated_hybrid_time: 6778530911018881024
read_time {
  read_ht: 6778530890413498368
  DEPRECATED_max_of_read_time_and_local_limit_ht: 6778530890413498368
  global_limit_ht: 6778530890413498368
  in_txn_limit_ht: 6778530911017603072
  local_limit_ht: 6778530890413498368
}
pgsql_batch {
  client: YQL_CLIENT_PGSQL
  stmt_id: 35948224
  schema_version: 0
  targets {
    tscall {
      opcode: 5
      operands {
        column_id: 0
      }
    }
  }
  column_refs {
    ids: 0
  }
  is_aggregate: true
  limit: 1024
  return_paging_state: true
  ysql_catalog_version: 1
  table_id: "000033e1000030008000000000004000"
}
proxy_uuid: ""
rejection_score: 0
batch_idx: 18446744073709551615
I0611 02:32:53.198460   159 service_util.cc:89] Check for tablet 8e32c8a62c4b4167abcc9f862bed812f peer a7ba92558e8848f580a9c29647bb78d5. Peer role is 1. Leader status is 4.
I0611 02:32:53.198473   159 hybrid_clock.cc:194] Current clock is higher than the last one. Resetting logical values. Time: { physical: 1654914773198474 }, Error: 500000
I0611 02:32:53.198483   159 mvcc.cc:584] T 8e32c8a62c4b4167abcc9f862bed812f P a7ba92558e8848f580a9c29647bb78d5: DoGetSafeTime, Now: { physical: 1654914773198474 }
I0611 02:32:53.198485   159 mvcc.cc:613] T 8e32c8a62c4b4167abcc9f862bed812f P a7ba92558e8848f580a9c29647bb78d5: DoGetSafeTime: ({ physical: 1654914768167358 }, { time: <invalid> lease: <max> }),  result = { physical: 1654914773198474 }
I0611 02:32:53.198556   159 tablet_service.cc:2277] Read time: { read: { physical: 1654914768167358 } local_limit: { physical: 1654914768167358 } global_limit: { physical: 1654914768167358 } in_txn_limit: { physical: 1654914773197657 } serial_no: 23 }, safe: { physical: 1654914773198474 }
I0611 02:32:53.198597   159 operation_counter.cc:84] [0x2d23f50] Update(1), result = 1
I0611 02:32:53.198654   159 bounded_rocksdb_iterator.cc:28] key_bounds_ = 0x0000000002d23e48 -> { lower: "" upper: "" }
I0611 02:32:53.198660   159 doc_rowwise_iterator.cc:1161] yb::Status yb::docdb::DocRowwiseIterator::DoInit(const T&) [with T = yb::docdb::DocPgsqlScanSpec] Seeking to DocKey([], [])
I0611 02:32:53.198670   159 pgsql_operation.cc:947] Started iterator
I0611 02:32:53.198854   159 pgsql_operation.cc:1002] Stopped iterator after 6 matches, 0 rows fetched
I0611 02:32:53.198858   159 pgsql_operation.cc:1004] Deadline is not exceeded
I0611 02:32:53.198880   159 operation_counter.cc:84] [0x2d23f50] Update(-1), result = 0
I0611 02:32:53.198894   159 hybrid_clock.cc:194] Current clock is higher than the last one. Resetting logical values. Time: { physical: 1654914773198894 }, Error: 500000
I0611 02:32:53.198907   159 yb_rpc.cc:406] Call yb.tserver.TabletServerService.Read 127.0.0.1:44212 => 127.0.0.1:9100 (request call id 76) took 0ms. Trace:
I0611 02:32:53.198909   159 yb_rpc.cc:407]
```

but then after that all of the logs were
```
I0611 02:32:54.153446   147 tablet_service.cc:2067] Received Read RPC: tablet_id: "8e32c8a62c4b4167abcc9f862bed812f"
include_trace: false
consistency_level: STRONG
propagated_hybrid_time: 6778530914930982912
pgsql_batch {
  client: YQL_CLIENT_PGSQL
  stmt_id: 35948416
  schema_version: 0
  targets {
    tscall {
      opcode: 5
      operands {
        column_id: 0
      }
    }
  }
  column_refs {
    ids: 0
  }
  is_aggregate: true
  limit: 1024
  return_paging_state: true
  ysql_catalog_version: 1
  table_id: "000033e1000030008000000000004000"
}
proxy_uuid: ""
rejection_score: 0
batch_idx: 18446744073709551615
I0611 02:32:54.153473   147 service_util.cc:89] Check for tablet 8e32c8a62c4b4167abcc9f862bed812f peer a7ba92558e8848f580a9c29647bb78d5. Peer role is 1. Leader status is 4.
I0611 02:32:54.153484   147 hybrid_clock.cc:194] Current clock is higher than the last one. Resetting logical values. Time: { physical: 1654914774153485 }, Error: 500000
I0611 02:32:54.153492   147 hybrid_clock.cc:194] Current clock is higher than the last one. Resetting logical values. Time: { physical: 1654914774153493 }, Error: 500000
I0611 02:32:54.153494   147 mvcc.cc:584] T 8e32c8a62c4b4167abcc9f862bed812f P a7ba92558e8848f580a9c29647bb78d5: DoGetSafeTime, Now: { physical: 1654914774153493 }
I0611 02:32:54.153496   147 mvcc.cc:613] T 8e32c8a62c4b4167abcc9f862bed812f P a7ba92558e8848f580a9c29647bb78d5: DoGetSafeTime: (<min>, { time: <invalid> lease: <max> }),  result = { physical: 1654914774153493 }
I0611 02:32:54.153501   147 hybrid_clock.cc:194] Current clock is higher than the last one. Resetting logical values. Time: { physical: 1654914774153501 }, Error: 500000
I0611 02:32:54.153503   147 tablet_service.cc:1982] Read time: { read: { physical: 1654914774153493 } local_limit: { physical: 1654914774153493 } global_limit: { physical: 1654914774653501 } in_txn_limit: <invalid> serial_no: 0 }
I0611 02:32:54.153510   147 tablet_service.cc:2277] Read time: { read: { physical: 1654914774153493 } local_limit: { physical: 1654914774153493 } global_limit: { physical: 1654914774653501 } in_txn_limit: <invalid> serial_no: 26 }, safe: { physical: 1654914774153493 }
I0611 02:32:54.153545   147 operation_counter.cc:84] [0x2d23f50] Update(1), result = 1
I0611 02:32:54.153616   147 bounded_rocksdb_iterator.cc:28] key_bounds_ = 0x0000000002d23e48 -> { lower: "" upper: "" }
I0611 02:32:54.153622   147 doc_rowwise_iterator.cc:1161] yb::Status yb::docdb::DocRowwiseIterator::DoInit(const T&) [with T = yb::docdb::DocPgsqlScanSpec] Seeking to DocKey([], [])
I0611 02:32:54.153636   147 pgsql_operation.cc:947] Started iterator
I0611 02:32:54.153694   147 pgsql_operation.cc:1002] Stopped iterator after 12 matches, 0 rows fetched
I0611 02:32:54.153698   147 pgsql_operation.cc:1004] Deadline is not exceeded
I0611 02:32:54.153710   147 operation_counter.cc:84] [0x2d23f50] Update(-1), result = 0
I0611 02:32:54.153745   147 hybrid_clock.cc:194] Current clock is higher than the last one. Resetting logical values. Time: { physical: 1654914774153746 }, Error: 500000
I0611 02:32:54.153774   147 yb_rpc.cc:406] Call yb.tserver.TabletServerService.Read 127.0.0.1:44212 => 127.0.0.1:9100 (request call id 77) took 0ms. Trace:
I0611 02:32:54.153777   147 yb_rpc.cc:407]
```

I'm not sure if the driver is doing something weird but it seems to be switching away from follower reads after the first query.

**Comment by jameshartig** [comment]
It looks like this can be fixed by changing the query to:

```go
err := readConn.QueryRow(context.Background(), "/*+ Set(transaction_read_only true) */ SELECT MAX(t1) FROM t", pgx.QuerySimpleProtocol(true)).Scan(&i)
```

The `pgx.QuerySimpleProtocol(true)` looks like it disables the automatic prepared query handling that pgx does. Should follower reads via the pg_hint not work with prepared queries?

**Comment by kmuthukk** [comment]
@jameshartig -- 

You ran into an issue with the extended query protocol + pg_hint that we too just noticed recently also. 

It's being tracked here:

https://github.com/yugabyte/yugabyte-db/issues/12741

and the fix is pretty much ready to land.. under final testing and review.

Nice work on your end as usual with providing a detailed explanation/test case.

**Comment by jameshartig** [comment]
Adding ` ysql_pg_conf_csv="log_min_messages=DEBUG5"` logs:
```
2022-06-11 03:10:08.858 UTC [308] DEBUG:  parse lrupsc_2_0: /*+ Set(transaction_read_only true) */ SELECT MAX(t1) FROM t
2022-06-11 03:10:08.858 UTC [308] DEBUG:  StartTransaction(1) name: unnamed; blockState: DEFAULT; state: INPROGRESS, ybDataSent: N, ybDataSentForCurrQuery: N, xid/subid/cid: 0/1/0
2022-06-11 03:10:08.876 UTC [308] DEBUG:  CommitTransaction(1) name: unnamed; blockState: STARTED; state: INPROGRESS, ybDataSent: N, ybDataSentForCurrQuery: N, xid/subid/cid: 0/1/0
2022-06-11 03:10:08.877 UTC [308] DEBUG:  bind <unnamed> to lrupsc_2_0
2022-06-11 03:10:08.877 UTC [308] DEBUG:  StartTransaction(1) name: unnamed; blockState: DEFAULT; state: INPROGRESS, ybDataSent: N, ybDataSentForCurrQuery: N, xid/subid/cid: 0/1/0
2022-06-11 03:10:08.900 UTC [308] DEBUG:  Doing read from followers
2022-06-11 03:10:08.901 UTC [308] DEBUG:  yb_exec_query_wrapper attempt 0 for /*+ Set(transaction_read_only true) */ SELECT MAX(t1) FROM t
2022-06-11 03:10:08.902 UTC [308] DEBUG:  CommitTransaction(1) name: unnamed; blockState: STARTED; state: INPROGRESS, ybDataSent: N, ybDataSentForCurrQuery: N, xid/subid/cid: 0/1/0
```
The first time and then
```
2022-06-11 03:10:09.855 UTC [308] DEBUG:  bind <unnamed> to lrupsc_2_0
2022-06-11 03:10:09.855 UTC [308] DEBUG:  StartTransaction(1) name: unnamed; blockState: DEFAULT; state: INPROGRESS, ybDataSent: N, ybDataSentForCurrQuery: N, xid/subid/cid: 0/1/0
2022-06-11 03:10:09.855 UTC [308] DEBUG:  Doing read from followers
2022-06-11 03:10:09.855 UTC [308] DEBUG:  yb_exec_query_wrapper attempt 0 for /*+ Set(transaction_read_only true) */ SELECT MAX(t1) FROM t
2022-06-11 03:10:09.857 UTC [308] DEBUG:  CommitTransaction(1) name: unnamed; blockState: STARTED; state: INPROGRESS, ybDataSent: N, ybDataSentForCurrQuery: N, xid/subid/cid: 0/1/0
```

Despite the second one having `consistency_level: STRONG`.

**Comment by jameshartig** [comment]
@kmuthukk Ah, just missed your reply! Good to know. Should I close this in favor of that issue then? Will that be backported to 2.12? Thanks again

**Comment by kmuthukk** [comment]
Yes - we will definitely backport to 2.12 also.

CC: @tanujnay112 , @amit-yb .

Closing this as a duplicate of https://github.com/yugabyte/yugabyte-db/issues/12741

**Comment by kmuthukk** [comment]
@jameshartig -- fix for #12741 has been committed, and backported to 2.12 and 2.14 branches. The next patch (3rd digit) release that comes out on those branches should contain the fix.

**Comment by jameshartig** [comment]
@kmuthukk thanks for the quick fix! I'll keep an eye out for the next release.
